src/TrackArtist.cpp

```diff
@@ -239,6 +240,35 @@ void TrackArt::DrawNegativeOffsetTrackArrows(
                 rect.x + 6, rect.y + rect.height - 12);
 }
 
+wxString TrackArt::TruncateText(wxDC& dc, const wxString& text, const int maxWidth)
+{
+   static const wxString ellipsis = "\u2026";
+
+   if (dc.GetTextExtent(text).GetWidth() <= maxWidth)
+       return text;
+
+   auto left = 0;
+   auto right = static_cast<int>(text.Length() - 2);
+
+   while (left <= right)
+   {
+      auto middle = (left + right) / 2;
+      auto str = text.SubString(0, middle).Trim() + ellipsis;
+      auto strWidth = dc.GetTextExtent(str).GetWidth();
+      if (strWidth < maxWidth)
+         left = middle + 1;
+      else if (strWidth > maxWidth)
+         right = middle - 1;
+      else
+         return str;
+   }
+   auto index = std::min(left, right);
+   if (index >= 0)
+      return text.SubString(0, index).Trim() + ellipsis;
```

Comment:
This might be simplified and clarified by using this standard function https://en.cppreference.com/w/cpp/algorithm/partition_point

Reply:
But i don't see how to use predicate in that case, if I understand you correctly you suggest this:
`std::partition_point(text.begin(), text.end(), [](auto char) { return /*how to get length?*/ }`

Reply:
I think you could take the char by reference and use its address.  A bit of a cheat though.

Reply:
If you don’t do that, I would prefer you clarify the loop by stating invariant conditions.  I think the invariant is: left - 1 is an allowed length for trimming the string or -1, and right +1 is a disallowed length.

If the loop terminates, the loop condition is false, so the min is right.  More, you can prove right == left - 1, and therefore a good length or -1 because the invariant is preserved.

The lack of comments a out the invariant and the unnecessary min makes me question how carefully you reasoned.

Reply:
>I think you could take the char by reference and use its address. A bit of a cheat though.

I would avoid making suggestions about internal representation of strings in the wxString, though it's been said that it uses std::basic_string as a container, encoding is another problem

>so the min is right

Right, thanks!